src/test_application/OSPMagnum.cpp

```diff
@@ -0,0 +1,209 @@
+#include <Magnum/PixelFormat.h>
+
+#include <iostream>
+
+
+OSPMagnum::OSPMagnum(const Magnum::Platform::Application::Arguments& arguments,
+                     osp::OSPApplication &ospApp) :
+        Magnum::Platform::Application{
+            arguments,
+            Configuration{}.setTitle("OSP-Magnum").setSize({1280, 720})},
+        m_userInput(12),
+        m_ospApp(ospApp)
+{
+    //.setWindowFlags(Configuration::WindowFlag::Hidden)
+
+    m_timeline.start();
+
+}
+
+void OSPMagnum::drawEvent()
+{
+
+    Magnum::GL::defaultFramebuffer.clear(Magnum::GL::FramebufferClear::Color
+                                         | Magnum::GL::FramebufferClear::Depth);
+
+//    if (m_area)
+//    {
+//        //Scene3D& scene = m_area->get_scene();
+//        //if (!m_area->is_loaded_active())
+//        //{
+//        //    // Enable active area if not already done so
+//        //    m_area->activate();
+//        //}
+
+//       // std::cout << "deltaTime: " << m_timeline.previousFrameDuration() << "\n";
+
+//        // TODO: physics update
+//        m_userInput.update_controls();
+//        m_area->update_physics(1.0f / 60.0f);
+//        m_userInput.clear_events();
+//        // end of physics update
+
+//        m_area->draw_gl();
+//    }
+
+    m_userInput.update_controls();
+
+    for (auto &[name, scene] : m_scenes)
+    {
+        scene.update();
+    }
+
+    m_userInput.clear_events();
+
+    for (auto &[name, scene] : m_scenes)
+    {
+        scene.update_hierarchy_transforms();
+
+
+        // temporary: draw using first camera component found
+        scene.draw(scene.get_registry().view<osp::active::ACompCamera>().front());
+    }
+
+
+    // TODO: GUI and stuff
+
+    swapBuffers();
+    m_timeline.nextFrame();
+    redraw();
+}
+
+
+
+void OSPMagnum::keyPressEvent(KeyEvent& event)
+{
+    if (event.isRepeated()) { return; }
+    m_userInput.event_raw(osp::sc_keyboard, (int) event.key(),
+                          osp::UserInputHandler::ButtonRawEvent::PRESSED);
+}
+
+void OSPMagnum::keyReleaseEvent(KeyEvent& event)
+{
+    if (event.isRepeated()) { return; }
+    m_userInput.event_raw(osp::sc_keyboard, (int) event.key(),
+                          osp::UserInputHandler::ButtonRawEvent::RELEASED);
+}
+
+void OSPMagnum::mousePressEvent(MouseEvent& event)
+{
+    m_userInput.event_raw(osp::sc_mouse, (int) event.button(),
+                          osp::UserInputHandler::ButtonRawEvent::PRESSED);
+}
+
+void OSPMagnum::mouseReleaseEvent(MouseEvent& event)
+{
+    m_userInput.event_raw(osp::sc_mouse, (int) event.button(),
+                          osp::UserInputHandler::ButtonRawEvent::RELEASED);
+}
+
+void OSPMagnum::mouseMoveEvent(MouseMoveEvent& event)
+{
+    m_userInput.mouse_delta(event.relativePosition());
+}
+
+void OSPMagnum::mouseScrollEvent(MouseScrollEvent & event)
+{
+    m_userInput.scroll_delta(static_cast<osp::Vector2i>(event.offset()));
+}
+
+osp::active::ActiveScene& OSPMagnum::scene_add(const std::string &name)
```

Comment:
There is `const std::string&` all over that can be replaced too. I'll keep this in mind for all future code.

Reply:
Sounds good

Reply:
This function has to convert the std::string_view to an std::string anyways, so it's probably not ideal for this function in particular.
```cpp
osp::active::ActiveScene& OSPMagnum::scene_add(const std::string &name)
{
    auto pair = m_scenes.try_emplace(name, m_userInput, m_ospApp);
    return pair.first->second;
}
```

Reply:
Ah, then this would be an "owning parameter usage".

I recommend making two versions of the function then.

One that accepts the string by const&, and the other that accepts it by && + std::move()

That allows the consumer of the function to get move semantics if they want them, which can allow the consumer to skip an allocation-and-copy that they always have to pay for with the const& version.

Reply:
Won't the same functionality be achieved with passing by value, and an std::move() into the emplace function? I'm also going to change the name of the function to "scene_create", as it's initializing a new scene.

```cpp
osp::active::ActiveScene& OSPMagnum::scene_create(std::string name)
{
    auto const& [it, success] = m_scenes.try_emplace(std::move(name), m_userInput, m_ospApp);
    return it->second;
}
```

Reply:
try_emplace implies that emplacement might fail.

If emplacement fails, I, as the consumer of the function, don't want to pay the price of allocating/copying the string parameter.

That's why I said the optimal way to do this is to have 2 functions. In my previous comment I said a const& and a && version, but I realized that std::string_view is a perfect super-set of the const& case, so I recommend that instead.

```
// The caller of the function doesn't have a std::string, so the std::string_view provides a way to call the function
// that only causes an allocation if emplacement will succeed.
osp::active::ActiveScene& OSPMagnum::scene_create(std::string_view name)
{
    auto const& [it, success] = m_scenes.try_emplace(name, m_userInput, m_ospApp);
    return it->second;
}

// In the case where we have an actual std::string, and we know we're done with it (because it's been std::moved() into the function), we can save an allocate and copy by providing this overload.
osp::active::ActiveScene& OSPMagnum::scene_create(std::string && name)
{
    auto const& [it, success] = m_scenes.try_emplace(std::move(name), m_userInput, m_ospApp);
    return it->second;
}
```